packages/flutter_tools/lib/src/device.dart

```diff
@@ -210,103 +275,133 @@ abstract class DeviceManager {
-      await refreshAllConnectedDevices(timeout: timeout);
-    }
+/// A class for determining how to filter devices based on if they are supported.
+///
+/// If [mustBeSupportedByFlutter] is true, only devices supported by Flutter
+/// will be included.
+///
+/// If [mustBeSupportedForProject] is true, only devices supported by Flutter
+/// and supported for the provided [flutterProject] will be included.
+///
+/// If [mustBeSupportedForAll] is true, only devices supported by Flutter,
+/// supported for the provided [flutterProject], and supported for
+/// `--device all` will be included.
+class DeviceDiscoverySupportFilter {
+  DeviceDiscoverySupportFilter({
+    required this.flutterProject,
+    this.mustBeSupportedByFlutter = false,
+    this.mustBeSupportedForProject = false,
+    this.mustBeSupportedForAll = false,
+  });
 
-    List<Device> devices = (await getDevices())
-        .where((Device device) => device.isSupported()).toList();
-
-    if (hasSpecifiedAllDevices) {
-      // User has specified `--device all`.
-      //
-      // Always remove web and fuchsia devices from `--all`. This setting
-      // currently requires devices to share a frontend_server and resident
-      // runner instance. Both web and fuchsia require differently configured
-      // compilers, and web requires an entirely different resident runner.
-      devices = <Device>[
-        for (final Device device in devices)
-          if (await device.targetPlatform != TargetPlatform.fuchsia_arm64 &&
-              await device.targetPlatform != TargetPlatform.fuchsia_x64 &&
-              await device.targetPlatform != TargetPlatform.web_javascript &&
-              isDeviceSupportedForProject(device, flutterProject))
-            device,
-      ];
-    } else if (!hasSpecifiedDeviceId) {
-      // User did not specify the device.
-
-      // Remove all devices which are not supported by the current application.
-      // For example, if there was no 'android' folder then don't attempt to
-      // launch with an Android device.
-      devices = <Device>[
-        for (final Device device in devices)
-          if (isDeviceSupportedForProject(device, flutterProject))
-            device,
-      ];
-
-      if (devices.length > 1) {
-        // If there are still multiple devices and the user did not specify to run
-        // all, then attempt to prioritize ephemeral devices. For example, if the
-        // user only typed 'flutter run' and both an Android device and desktop
-        // device are available, choose the Android device.
-
-        // Ephemeral is nullable for device types where this is not well
-        // defined.
-        final List<Device> ephemeralDevices = <Device>[
-          for (final Device device in devices)
-            if (device.ephemeral == true)
-              device,
-        ];
-
-        if (ephemeralDevices.length == 1) {
-          devices = ephemeralDevices;
-        }
-      }
+  final FlutterProject? flutterProject;
```

Comment:
can this really be null?

Reply:
Yes, when `includeUnsupportedDevices` is true, the `flutterProject` will be null. This is true in `flutter logs` and can be true in `flutter drive`.
https://github.com/flutter/flutter/blob/455cc93680158669b3208e816a7779f011bebc18/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1526

https://github.com/flutter/flutter/blob/455cc93680158669b3208e816a7779f011bebc18/packages/flutter_tools/lib/src/commands/logs.dart#L40

https://github.com/flutter/flutter/blob/455cc93680158669b3208e816a7779f011bebc18/packages/flutter_tools/lib/src/commands/drive.dart#L211